libbpf-async/src/ringbuf.rs

```diff
@@ -99,31 +100,36 @@ impl AsyncRead for RingBuffer {
 
                 if (len & BPF_RINGBUF_BUSY_BIT) == 0 {
                     cons_pos += RingBuffer::roundup_len(len) as u64;
-                    let sample = unsafe {
-                        std::slice::from_raw_parts_mut(
-                            len_ptr.offset(BPF_RINGBUF_HDR_SZ as isize) as *mut u8,
-                            len as usize,
-                        )
-                    };
-                    unsafe {
-                        let b = &mut *(buf.unfilled_mut() as *mut [std::mem::MaybeUninit<u8>]
-                            as *mut [u8]);
-                        len = std::cmp::min(len, buf.capacity() as u32);
-                        for i in 0..len {
-                            b[i as usize] = sample[i as usize];
+                    if (len & BPF_RINGBUF_DISCARD_BIT) == 0 {
+                        let sample = unsafe {
+                            std::slice::from_raw_parts_mut(
+                                len_ptr.offset(BPF_RINGBUF_HDR_SZ as isize) as *mut u8,
+                                len as usize,
+                            )
+                        };
+                        unsafe {
+                            let b = &mut *(buf.unfilled_mut() as *mut [std::mem::MaybeUninit<u8>]
+                                as *mut [u8]);
+                            len = std::cmp::min(len, buf.capacity() as u32);
+                            for i in 0..len {
+                                b[i as usize] = sample[i as usize];
+                            }
+                            buf.assume_init(len as usize);
+                            buf.advance(len as usize);
                         }
-                        buf.assume_init(len as usize);
-                        buf.advance(len as usize);
                     }
-
                     std::sync::atomic::fence(std::sync::atomic::Ordering::SeqCst);
                     unsafe {
                         std::ptr::write_volatile(
                             self.consumer as *mut std::os::raw::c_ulong,
                             cons_pos,
                         )
                     };
-                    return Poll::Ready(Ok(()));
+                    if (len & BPF_RINGBUF_DISCARD_BIT) == 0 {
+                        return Poll::Ready(Ok(()));
+                    } else {
+                        continue;
```

Comment:
`continue` is necessary?

Reply:
I _suspect_ it is. 
My reasoning was that without the `continue`, if there is a discard (without a wakeup) followed by a commit (with a wakeup), then the rust-side would just move the ringbuf consumer pointer forward and go back to sleep. This instead of immediately yielding the the fresh data? I could be completely wrong about this and just misunderstanding what the `ev.clear_ready()` does.
I am unfortunately no authority on async or tokio.

Reply:
Understood. We shouldn't call _self.async_fd.poll_read_ready(cx)_ until processing all the available entries, I guess.